### ISSUE TITLE
Allow VRL programs to emit multiple events

### ISSUE THREAD

**Issue by lucperkins** [timberio/vector#6330]
A user came on Discord and requested functionality along these lines: https://discuss.elastic.co/t/how-to-split-an-event-to-multiple-events/162406/3. I recommended the Lua transform as a potential solution in that Lua enables you to call `emit` multiple times. But allowing for this in VRL would take a chunk out of the range of use cases only covered by Lua/Wasm.

To reformulate the request in simpler terms, take this JSON log:

```json
{"stooges": ["Larry", "Moe", "Curly"]}
```

The request would be to be able to produce three separate events from this input: `{"stooge": "Larry"}`, `{"stooge": "Moe"}`, and `{"stooge": "Curly"}`. Imagine something like this in VRL:

```ruby
# Event 1
.stooge = del(.stooges[0])
emit(.)

# Event 2
.stooge = del(.stooges[1])
emit(.)

# Event 3
.stooge = del(.stooges[2])
emit(.)
```

Or to use more elegant pseudocode in the absence of a completed `for` construct:

```ruby
for stooge in .stooges
  .stooge = stooge
  emit(.)
end
```

This would surely go against the grain of how we've built and conceived of VRL thus far, of course, in that it violates the "one event in, one event out" principle. But IMHO it's not so far out there that we shouldn't at least formulate principled reasons for excluding this possibility.

**Comment by JeanMertz** [comment]
On a technical level, the way I'd see this working is like this.

**First**, we currently have a `Target` trait which gets implemented by `vector::Event`:

```rust
pub trait Target {
    fn insert(&mut self, path: &Path, value: Value) -> Result<(), String>;
    fn get(&self, path: &Path) -> Result<Option<Value>, String>;
    fn remove(&mut self, path: &Path, compact: bool) -> Result<Option<Value>, String>;
}
```

To support this, we'd add an extra function to this list:

```rust
fn emit(&self, value: Value) -> Result<(), String>;
```

This instructs Vector (or whoever else uses VRL) to emit the event and report back if it succeeded or not.

**Second**, we'd introduce a new `emit` function into VRL which wraps this trait call. It would be a fallible function, with the following signature:

```ruby
emit[!](value: Object)
```

**Third**, then it's just a matter of wiring up Vector to actually emit the event as expected, and that's it. 

---

On a conceptual level, there's still some things to consider.

1. As you mentioned, this is getting into territory we initially wanted to avoid. Technically, it's easy to add as shown above, but it might make future changes to the language more complex (although, as shown, there are no core language changes needed for this to work, so I think the likelihood of this happening is relatively slim).

2. Separate from the technical question, it might make using VRL too confusing, although again, it's "just" a function, and you can ignore it if you have no need for it.

3. Lastly, the one confusing part about this I could see is one you made in your example, namely, your example will emit _four_ events, not three because after the program finishes, whatever the state `.` is in, will also be emitted. If we don't want that, _then_ things become more complex because we'd have to either (a) detect if `emit` is used, and instruct vector not emit the final event, or (b) add a configuration flag to the remap transform to ignore the final event, forcing you (potentially with a compile-time check) to use `emit` at least once.

I think this last point is the biggest reason why we might not want to do this, but I do admit the simplicity of adding this, and the potential use-cases it unlocks are tempting.

But maybe that temptation should be smacked down by @binarylogic laser-focus on usability and simplicity for VRL.

**Comment by JeanMertz** [comment]
Another potential pitfall:

Using `emit(.)` would emit the current state of the target, but it'll leave the existing target as is, so you don't get a "clean slate" target as `.` to work with. If you call `emit(.)` again without making any changes to the target, you'll essentially send a duplicate event.

This seems logical to me, but it might be confusing to others, maybe?

**Comment by jszwedko** [comment]
Another user request from discord: https://discord.com/channels/742820443487993987/764187584452493323/808744293945704479

**Comment by jszwedko** [comment]
Also of note is that emitting multiple events is generally necessary to deal with the output of `parse_aws_cloudwatch_log_subscription_message` as the output of that is an array of events.

This is actually making me realize maybe there is another path here? Could we treat `.` being set to an `Array` type as indicating that multiple events should be generated? This would let us write scripts like:

```
subscription_message = (parse_aws_cloudwatch_log_subscription_message(.message)
# in reality, we'd probably also want to map the log_events to add additional metadata
. = subscription_message.log_events
```

**Comment by lukesteensen** [comment]
I'm in favor of this capability in general, but I agree there are a lot of pitfalls and that we shouldn't sacrifice VRL's relatively simple mental model (i.e. there's one input event at `.`, you manipulate it, it gets returned).

We could take some inspiration from the Lua transform v1/v2 changes, where the context in which the script is run became more explicit and users got a couple of different "mode" options. I don't think we'd want to go quite as far, since again we want to keep the simple use cases as simple as possible, but we could consider some slightly more explicit mode that users can opt into to get more features like this.

**Comment by lukesteensen** [comment]
It's also interesting to compare to `jq`, where this example is just `.stooges[] | { stooge: . }`. Their more pipeline-like execution context pays dividends there.

SQL's `unnest` array function could be another one to consider as well.

**Comment by jszwedko** [comment]
This is closed by #7404 and #7267
